### PR TITLE
nvidia-container-toolkit: add "nvidia" to services.xserver.videoDrivers

### PR DESCRIPTION
**Pull request by ereslibre**
This will ensure that all requirements are met when the nvidia-container-toolkit is enabled in order to generate the CDI specs for the machine.

Also, if `config.virtualisation.docker.rootless.enable` is true, automatically enable the CDI feature if the Docker version > 25 for the rootless installation.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/services/hardware/nvidia-container-toolkit/default.nix [hunk 1]
```diff
@@ -77,50 +77,63 @@
...
+        (lib.mkIf config.hardware.nvidia-container-toolkit.mount-nvidia-docker-1-directories
+          [{ hostPath = "${lib.getLib nvidia-driver}/lib";
+             containerPath = "/usr/local/nvidia/lib"; }
+           { hostPath = "${lib.getLib nvidia-driver}/lib";
+             containerPath = "/usr/local/nvidia/lib64"; }])
+      ]);
+    };
+
+    services.xserver.videoDrivers = lib.mkIf
+      (!config.hardware.nvidia.datacenter.enable) [ "nvidia" ];
```
Comment (SomeoneSerge) [inline nixos/modules/services/hardware/nvidia-container-toolkit/default.nix:134]: I don't think services.xserver.videoDrivers feature any deduplication logic so a user specifying a driver explicitly would likely end up with `videoDrivers = [ "nvidia" "nvidia" ]`. Should make any difference to the symlink farms though

File: nixos/modules/services/hardware/nvidia-container-toolkit/default.nix [hunk 2]
```diff
@@ -77,50 +77,63 @@
...
+
+      rootless.daemon.settings = lib.mkIf
+        (config.virtualisation.docker.rootless.enable &&
+         (lib.versionAtLeast config.virtualisation.docker.package.version "25")) {
+           features.cdi = true;
+         };
+    };
+
+    hardware = {
+      graphics.enable = lib.mkIf (!config.hardware.nvidia.datacenter.enable) true;
```
Comment (SomeoneSerge) [inline nixos/modules/services/hardware/nvidia-container-toolkit/default.nix:95]: I think this one should be enabled regardless? So as to generate `/run/opengl-driver/lib`

**Comment by SomeoneSerge** [comment]
Comments here and in https://github.com/NixOS/nixpkgs/pull/344188 may be  addressed in a follow-up PR

**Comment by phanirithvij** [comment]
From a git bisect session, I found this broke my x11 graphical session (kde plasma6), due to adding "nvidia" to the service.xserver.videoDrivers.

I never messed with nvidia settings till recently with `hardware.nvidia.open = false` and I have been using what came with the initial configuration.nix. i.e. default of `[ "modesetting" "fbdev" ]`.

So for some people this could be a breaking change, so mention this on [this thread](https://discourse.nixos.org/t/breaking-changes-announcement-for-unstable)?

**Comment by ereslibre** [comment]
Hello @phanirithvij and sorry for the breakage!

I am opening a PR to fix this issue. No need to add that as part of breaking changes as I understand because it’s a bug. The “nvidia” string should not be appended to videoDrivers if hardware.nvidia.open is true.

**Comment by SomeoneSerge** [comment]
@phanirithvij can you confirm, are you saying `hardware.nvidia-toolkit-enable.enable` was `false` and yet this module added `"nvidia"` to `videoDrivers`?

**Comment by ereslibre** [comment]
@phanirithvij @SomeoneSerge as far as I understand, this assertion should be enough: https://github.com/NixOS/nixpkgs/pull/350339.

@phanirithvij can you confirm that you tried to `hardware.nvidia-toolkit-enable.enable = true;` and `hardware.nvidia.open = true;` as well? Just in case I misunderstood.

**Comment by phanirithvij** [comment]
> you tried to hardware.nvidia-toolkit-enable.enable = true; and hardware.nvidia.open = true; as well? Just in case I misunderstood.

No I can't use open source nvidia drivers. I have it set as `false`

`hardware.nvidia.open = true` doesn't work on my laptop (with intel+nvidia integrated graphics) at all

```console
$ journalctl -rb -p 0..5
Oct 22 08:29:52 iron (udev-worker)[897]: nvidia: Process '/nix/store/717iy55ncqs0wmhdkwc5fg2vci5wbmq8-bash-5.2p32/bin/bash -c 'mknod -m 666 /dev/nvidiactl c 195 255'' failed with exit code 1.
Oct 22 08:29:52 iron (udev-worker)[897]: nvidia: Process '/nix/store/717iy55ncqs0wmhdkwc5fg2vci5wbmq8-bash-5.2p32/bin/bash -c 'mknod -m 666 /dev/nvidiactl c 195 255'' failed with exit code 1.
Oct 22 08:29:52 iron kernel: NVRM: None of the NVIDIA devices were initialized.
Oct 22 08:29:52 iron kernel: NVRM: The NVIDIA probe routine failed for 1 device(s).
Oct 22 08:29:52 iron kernel: nvidia: probe of 0000:01:00.0 failed with error -1
Oct 22 08:29:52 iron kernel: NVRM: The NVIDIA GPU 0000:01:00.0 (PCI ID: 10de:134d)
                             NVRM: installed in this system is not supported by open
                             NVRM: nvidia.ko because it does not include the required GPU
                             NVRM: System Processor (GSP).
                             NVRM: Please see the 'Open Linux Kernel Modules' and 'GSP
                             NVRM: Firmware' sections in the driver README, available on
                             NVRM: the Linux graphics driver download page at
                             NVRM: www.nvidia.com.
```

the following is for when I have **hardware.nvidia.open = false;**

```console
$ nix-shell -p inxi pciutils virtualgl --command 'inxi -G'
Graphics:
  Device-1: Intel UHD Graphics 620 driver: i915 v: kernel
  Device-2: NVIDIA GM108M [GeForce 940MX] driver: nouveau v: kernel
  Display: x11 server: X.org v: 1.21.1.13 with: Xwayland v: 24.1.3 driver:
    X: loaded: modesetting dri: iris,nouveau gpu: i915
    resolution: 1920x1080~60Hz
  API: EGL Message: No EGL data available.
  API: OpenGL v: 4.6 vendor: intel mesa v: 24.2.4 renderer: Mesa Intel UHD
    Graphics 620 (KBL GT2)
```

and `services.xserver.videoDrivers = [ "nvidia" ]` doesn't let me start x11 either.

I've just realised, after switching back to before this pr, nvidia-container-toolkit was never working in the first place. (I just enabled it and left it there sitting in my config)

```console
$ systemctl status nvidia-container-toolkit-cdi-generator.service
● nvidia-container-toolkit-cdi-generator.service - Container Device Interface (CDI) for Nvidia generator
     Loaded: loaded (/etc/systemd/system/nvidia-container-toolkit-cdi-generator.service; enabled; preset: ignored)
     Active: active (exited) since Tue 2024-10-22 09:17:03 IST; 21min ago
 Invocation: 30315690a8244d598da69e0bbfd8a9f9
    Process: 993 ExecStart=/nix/store/dj6j7ipi2dnp167gdi285asylmbjz87w-nvidia-cdi-generator/bin/nvidia-cdi-generator (code=exited, status=0/SUCCESS)
   Main PID: 993 (code=exited, status=0/SUCCESS)
         IP: 0B in, 0B out
   Mem peak: 23.6M
        CPU: 75ms

Oct 22 09:17:03 iron systemd[1]: Starting Container Device Interface (CDI) for Nvidia generator...
Oct 22 09:17:03 iron nvidia-cdi-generator[1013]: time="2024-10-22T09:17:03+05:30" level=info msg="Auto-detected mode as \"nvml\""
Oct 22 09:17:03 iron nvidia-cdi-generator[1013]: time="2024-10-22T09:17:03+05:30" level=error msg="failed to generate CDI spec: failed to create device CDI specs: failed to initialize NVML: ERROR_LIBRARY_NOT_FOUND"
Oct 22 09:17:03 iron systemd[1]: Finished Container Device Interface (CDI) for Nvidia generator.
```

I guess you don't have to worry, since it looks like my gpu is not supported?

**Comment by ereslibre** [comment]
> I guess you don't have to worry, since it looks like my gpu is not supported? so I am happy with disabling nvidia-container-toolkit

As far as the nvidia-container-toolkit is concerned, with https://github.com/NixOS/nixpkgs/pull/350339 at least now we'll notify users that it won't work unless the nvidia driver is present, given it was silently failing in your setup. Again, sorry for the graphical environment breakage!

**Comment by phanirithvij** [comment]
Ok I got it with `services.xserver.videoDrivers = [ "modesetting" "nvidia" ]`, nvidia-container-toolkit works as well now.

> sorry for the graphical environment breakage

It's okay, I am running unstable for this reason, catching things early.

**Pull request merged**